File: Data Structures & Algorithms/max-water-container/submission-3.java
class Solution {
    public int maxArea(int[] heights) {
        
        int l = 0, r = heights.length - 1;
        int ans = Integer.MIN_VALUE;
        while(l < r)
        {
            int h = Math.min(heights[l],heights[r]);
            int w = r-l;

            int area = h * w;
            ans = Math.max(ans,area);

            if(heights[l] < heights[r])
                l++;
            else
                r--;
        }

        return ans;
    }
}

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-3.py
class Solution:

    def canShip(self,weights: List[int],cap: int, totalShips: int) -> boolean:

        shipCount,currentCap = 1,cap
        for i in range(0,len(weights)):
            if currentCap - weights[i] < 0:
                shipCount += 1
                if shipCount > totalShips:
                    return False
                currentCap = cap
            currentCap -= weights[i]
        
        return True

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low,high,n = float('-inf'),0,len(weights)

        for i in range(0,n):
            low = max(low,weights[i])
            high += weights[i]
        
        ans = high

        while low <= high:
            cap = low + (high - low) // 2
            if self.canShip(weights,cap,days):
                ans = min(ans,cap)
                high = cap - 1
            else:
                low = cap + 1

        return ans